### ISSUE TITLE
Extend UnusedLocalVariableCheck to support pattern variables

### ISSUE THREAD

**Issue by mahfouz72** [checkstyle/checkstyle#15085]
child of #14942

we should extend `UnusedLocalVariable` to support pattern variables. The unused pattern variables should be unnamed patterns or unnamed pattern variables. This Improves the readability of record patterns by eliding unnecessary nested type patterns. The main goal is to ensure that unused pattern variables are detected and violated, prompting developers to either use them or declare them as unnamed variables.

To be clear a pattern is a local variable.
Reference: https://docs.oracle.com/javase/specs/jls/se17/html/jls-14.html#jls-14.30:~:text=A%20local%20variable%20declaration%20can%20appear%20in%20the%20following%20locations%3A

> A local variable declaration can appear in the following locations:

> - a local variable declaration statement in a block ([§14.4.2](https://docs.oracle.com/javase/specs/jls/se17/html/jls-14.html#jls-14.4.2))

> - the header of a basic for statement ([§14.14.1](https://docs.oracle.com/javase/specs/jls/se17/html/jls-14.html#jls-14.14.1))
> 
> - the header of an enhanced for statement ([§14.14.2](https://docs.oracle.com/javase/specs/jls/se17/html/jls-14.html#jls-14.14.2))

> - the resource specification of a try-with-resources statement ([§14.20.3](https://docs.oracle.com/javase/specs/jls/se17/html/jls-14.html#jls-14.20.3))

> - a pattern ([§14.30.1](https://docs.oracle.com/javase/specs/jls/se17/html/jls-14.html#jls-14.30.1))

---
Examples

1. Switch and case labels

```Java
sealed abstract class Ball permits RedBall, GreenBall { }
final  class RedBall   extends Ball { }
final  class GreenBall extends Ball { }
record Box<T extends Ball>(T content) { }

public class Test {
    public void PatternVariableInCaseLabels(Ball ball) {
        switch (ball) {
            case RedBall redBall -> process(ball); // violation, unused variable 'redBall'
            case GreenBall greenBall -> stopProcessing(ball);  // violation, unused variable 'greenBall'
        }
        // after fix
        switch (ball) {
            case RedBall _ -> process(ball);
            case GreenBall _ -> stopProcessing(ball);
        }
    }

    public void PatternVariableRecordPatternCaseLabels(Box<? extends Ball> box) {
        switch (box) {
            case Box(RedBall redBall) -> process(box); // violation, unused variable 'redBall'
            case Box(GreenBall greenBall) -> stopProcessing(box);  // violation, unused variable 'greenBall'
        }
        switch (box) {
            case Box(RedBall _) -> process(box); // violation, unused variable 'redBall'
            case Box(GreenBall _) -> stopProcessing(box);  // violation, unused variable 'greenBall'
        }

    }
    void process(Object obj) {}
    void stopProcessing(Object obj) {}
}
```

2. Contitional Statments

```Java
public class Test {

    public void PatternVariableConditionalStatements(Ball ball) {
        if (ball instanceof RedBall redBall) { // violation, unused variable 'redBall'
            process(ball);
        } else if (ball instanceof GreenBall greenBall) { // violation, unused variable 'greenBall'
            stopProcessing(ball);
        }
        // fix
        if (ball instanceof RedBall _) {
            process(ball);
        } else if (ball instanceof GreenBall _) {
            stopProcessing(ball);
        }
    }

    public void PatternVariableConditionalStatements2(Ball ball) {
        if (!(ball instanceof RedBall redBall)) {
            process(ball);  // red ball is not in this scope
        } else {
             process(redBall);  // ok red ball is used
        }
    }

    public void PatternVariableInRecordPatternInConditionalStatements(Box<? extends Ball> box) {
        if (box instanceof Box(RedBall redBall)) { // violation, unused variable 'redBall'
            process(box);
        } else if (box instanceof Box(GreenBall greenBall)) { // violation, unused variable 'greenBall'
            stopProcessing(box);
        }
        // fix
        if (box instanceof Box(RedBall _)) {
            process(box);
        } else if (box instanceof Box(GreenBall _)) {
            stopProcessing(box);
        }
       
     public void PatternVariableNestedRecordPatterns(Object o) {
         if (o instanceof Rectangle(ColoredPoint(int p, int x, String c), ColoredPoint lr)) {
             System.out.println(p + "" + x + c);
         }
         // fix
         if (o instanceof Rectangle(ColoredPoint(int p, int x, String c), ColoredPoint _)) {
             System.out.println(p + "" + x + c);
         }
         // another fix
         if (o instanceof Rectangle(ColoredPoint(int p, int x, String c), _)) {
             System.out.println(p + "" + x + c);
         }
      }
    }
}
```

---

pattern variables within conditional statements are subject to flow sensitivity. This means that the scope of these variables depend on the control flow of the program.

Reference for the pattern variable scope:

- [6.3.1. Scope for Pattern Variables in Expressions](https://docs.oracle.com/javase/specs/jls/se22/html/jls-6.html#jls-6.3.1)
- [6.3.2. Scope for Pattern Variables in Statements](https://docs.oracle.com/javase/specs/jls/se22/html/jls-6.html#jls-6.3.2)
- [6.3.3. Scope for Pattern Variables in case Labels](https://docs.oracle.com/javase/specs/jls/se22/html/jls-6.html#jls-6.3.3)

**Comment by nrmancuso** [comment]
@mahfouz72 it is not clear to me how this check currently works now on the input files, and how the check would change. Can you add some more information to the issue description?

**Comment by nrmancuso** [comment]
@mahfouz72 ping

**Comment by nrmancuso** [comment]
It might be good to break this into at least two issues (switch and instanceof). That said, we will not have time to complete this during GSOC so I am removing it from the project.

**Comment by ZaheerAhmadDev** [comment]
@romani Please approved the issue.

**Comment by mahfouz72** [comment]
@ZaheerAhmadDev Can you post the issue details here in the comments using our issue template, and show how the latest Checkstyle version handles all the code examples from the original post? I will review and approve it afterward.

**Comment by ZaheerAhmadDev** [comment]
For Switch and case labels:

```sealed abstract class Ball permits RedBall, GreenBall { }
final  class RedBall   extends Ball { }
final  class GreenBall extends Ball { }
record Box<T extends Ball>(T content) { }

public class Test {
    public void PatternVariableInCaseLabels(Ball ball) {
        switch (ball) {
            case RedBall redBall -> process(ball); // expected violation, unused variable 'redBall'
            case GreenBall greenBall -> stopProcessing(ball);  // expected violation, unused variable 'greenBall'
        }
        // after fix
        switch (ball) {
            case RedBall _ -> process(ball);
            case GreenBall _ -> stopProcessing(ball);
        }
    }

    public void PatternVariableRecordPatternCaseLabels(Box<? extends Ball> box) {
        switch (box) {
            case Box(RedBall redBall) -> process(box); // expected violation, unused variable 'redBall'
            case Box(GreenBall greenBall) -> stopProcessing(box);  // expected violation, unused variable 'greenBall'
        }
        switch (box) {
            case Box(RedBall _) -> process(box);
            case Box(GreenBall _) -> stopProcessing(box);
        }

    }
    void process(Object obj) {}
    void stopProcessing(Object obj) {}
}
```

 ```C:\Users\PMYLS\Desktop\pattern>type config.xml
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
  <module name="TreeWalker">
    <module name="UnusedLocalVariable"/>
  </module>
</module>

C:\Users\PMYLS\Desktop\pattern>java  -jar checkstyle-13.4.2-all.jar -c config.xml Test.java
Starting audit...
Audit done.
 ```

for Conditional Statements:
```
sealed abstract class Ball permits RedBall, GreenBall { }
final  class RedBall   extends Ball { }
final  class GreenBall extends Ball { }
record Box<T extends Ball>(T content) { }
record ColoredPoint(int x, int y, String color) { }
record Rectangle(ColoredPoint upperLeft, ColoredPoint lowerRight) { }


public class Test {

    public void PatternVariableConditionalStatements(Ball ball) {
        if (ball instanceof RedBall redBall) { // expected violation, unused variable 'redBall'
            process(ball);
        } else if (ball instanceof GreenBall greenBall) { // expected violation, unused variable 'greenBall'
            stopProcessing(ball);
        }
        // fix
        if (ball instanceof RedBall _) {
            process(ball);
        } else if (ball instanceof GreenBall _) {
            stopProcessing(ball);
        }
    }

    public void PatternVariableConditionalStatements2(Ball ball) {
        if (!(ball instanceof RedBall redBall)) {
            process(ball);  // red ball is not in this scope
        } else {
             process(redBall);  // ok red ball is used
        }
    }

    public void PatternVariableInRecordPatternInConditionalStatements(Box<? extends Ball> box) {
        if (box instanceof Box(RedBall redBall)) { // expected violation, unused variable 'redBall'
            process(box);
        } else if (box instanceof Box(GreenBall greenBall)) { // expected violation, unused variable 'greenBall'
            stopProcessing(box);
        }
        // fix
        if (box instanceof Box(RedBall _)) {
            process(box);
        } else if (box instanceof Box(GreenBall _)) {
            stopProcessing(box);
        }
    }

    public void PatternVariableNestedRecordPatterns(Object o) {
        if (o instanceof Rectangle(ColoredPoint(int p, int x, String c), ColoredPoint lr)) {
            System.out.println(p + "" + x + c);
        }
        // fix
        if (o instanceof Rectangle(ColoredPoint(int p, int x, String c), ColoredPoint _)) {
            System.out.println(p + "" + x + c);
        }
        // another fix
        if (o instanceof Rectangle(ColoredPoint(int p, int x, String c), _)) {
            System.out.println(p + "" + x + c);
        }
    }
}
```
 ```
C:\Users\PMYLS\Desktop\pattern>type config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
  <module name="TreeWalker">
    <module name="UnusedLocalVariable"/>
  </module>
</module>

C:\Users\PMYLS\Desktop\pattern>java  -jar checkstyle-13.4.2-all.jar -c config.xml src/Test.java
Starting audit...
Audit done.
 ```

**Comment by romani** [comment]
fix is merged

**Comment by romani** [comment]
@ZaheerAhmadDev, thanks a lot for valuable contribution !!!
